src/Faithlife.Tracing/ITraceFactory.cs

```diff
@@ -0,0 +1,10 @@
+using System;
+
+namespace Faithlife.Tracing
+{
+	public interface ITraceFactory
+	{
+		ITrace ExtractTrace(Func<string, string> extractKey);
+		ITrace CreateTrace(ITrace parent, TraceKind traceKind, string serviceName, string rpc, params string[] tags);
```

Comment:
`StartTrace` ?

Reply:
Extension methods might be nice as well, e.g. for a server trace with no parent and no "rpc":

```
public static ITrace StartServerTrace(this ITraceFactory traceFactory, string serviceName, params (string, string)[] tags)
```

Reply:
Seems like `kind` would be fine (vs. `traceKind`).

I don't yet understand `serviceName` and `rpc`.

Reply:
The interface method should really take an object (e.g. `TraceInfo`) so that properties can be added without breaking implementing types.

Reply:
That's a good point; I'm a little wary of creating extra allocations in an API that will be used on every request.

This needs more thought, but it might be a no-op to call `CreateTrace` when `parent == null`; the library could interpret this as "the current request isn't being sampled, ignore it" and return `null`, and the cycle repeats itself. It would be suboptimal to incur allocations just to make the method call. We could specify that the implementation MAY NOT cache the `TraceInfo` object, which would allow highly-sensitive callers to have a per-thread cache of `ThreadInfo` objects that they reuse.

Reply:
FWIW, the `params` array is already an allocation. It seems unlikely that one more allocation is going to be noticeable, and forward compatibility is pretty important.

Prohibiting caching the `TraceInfo` seems reasonable. If it's mutable, hopefully that would also be "obvious".

A related idea would be to use an interface (`ITraceInfo`), which highly-sensitive callers could leverage, for example by calculating tags just in time, if asked for.